components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt

```diff
@@ -0,0 +1,231 @@
+
+        val contentType = response.headers[CONTENT_TYPE]
+        val contentLength = response.headers[CONTENT_LENGTH]?.toLongOrNull()
+
+        addCompletedDownload(
+            title = filename,
+            description = filename,
+            isMediaScannerScannable = true,
+            mimeType = contentType ?: download.contentType ?: "*/*",
+            path = file.absolutePath,
+            length = contentLength ?: download.contentLength ?: file.length(),
+            showNotification = true, // TODO show when our notifications are blocked
+            uri = download.url.toUri(),
+            referer = download.referrerUrl?.toUri()
+        )
+        // }
+    }
+
+    @TargetApi(Build.VERSION_CODES.P)
+    @Suppress("Deprecation", "LongParameterList", "LongMethod")
+    private fun addCompletedDownload(
+        title: String,
+        description: String,
+        isMediaScannerScannable: Boolean,
+        mimeType: String,
+        path: String,
+        length: Long,
+        showNotification: Boolean,
+        uri: Uri,
+        referer: Uri?
+    ) = getSystemService<SystemDownloadManager>()!!.run {
+        if (SDK_INT >= Build.VERSION_CODES.N) {
+            addCompletedDownload(
+                title,
+                description,
+                isMediaScannerScannable,
+                mimeType,
+                path,
+                length,
+                showNotification,
+                uri,
+                referer
+            )
+        } else {
+            addCompletedDownload(
+                title,
+                description,
+                isMediaScannerScannable,
+                mimeType,
+                path,
+                length,
+                showNotification
+            )
+        }
+    }
+
+    private fun buildNotification(): Notification {
+        val channelId = DownloadNotificationChannel.ensureChannelExists(this)
+
+        return NotificationCompat.Builder(this, channelId)
+            .setSmallIcon(android.R.drawable.stat_sys_download)
+            .setContentTitle("Downloading...")
+            .setCategory(NotificationCompat.CATEGORY_PROGRESS)
+            .setProgress(1, 0, true)
+            .setOngoing(true)
+            .build()
+    }
+
+    private fun updateNotification() {
+        NotificationManagerCompat.from(this)
+            .notify(id(this), buildNotification())
+    }
+
+    companion object {
+        private const val ONGOING_DOWNLOAD_NOTIFICATION_TAG = "OngoingDownload"
+
+        private fun id(context: Context) =
+            NotificationIds.getIdForTag(context, ONGOING_DOWNLOAD_NOTIFICATION_TAG)
+    }
+}
+
+internal object DownloadNotificationChannel {
+    private const val NOTIFICATION_CHANNEL_ID = "Downloads"
+
+    /**
+     * Check if notifications have been blocked for this channel or its group.
+     */
+    fun isBlocked(context: Context) =
+        NotificationUtils.isNotificationBlocked(context, NOTIFICATION_CHANNEL_ID)
```

Comment:
I can't find where this is actually used? 🤔 - Do we even need to check for that?

Reply:
I was initially using this for the completion notification, falling back to native if the channel is blocked. I ended up not implementing a download complete notification for now, so I can pull this code out into its own PR.